File: old/core/aggregators/reddit/urls.py
"""Reddit URL utilities."""

import html
import logging
import re
from typing import Any, Dict, Optional

import prawcore.exceptions

from .auth import get_praw_instance

logger = logging.getLogger(__name__)

REDDIT_API_BASE = "https://www.reddit.com"


def decode_html_entities_in_url(url: str) -> str:
    """
    Decode HTML entities in URLs using html.unescape.

    Args:
        url: URL string

    Returns:
        Decoded URL
    """
    return html.unescape(url)


def fix_reddit_media_url(url: Optional[str]) -> Optional[str]:
    """
    Fix Reddit media URLs by decoding HTML entities.
    Reddit API often returns URLs with double-escaped entities like &amp;amp;.

    Args:
        url: URL from Reddit API

    Returns:
        Fixed URL or None
    """
    if not url:
        return None

    # Use html.unescape twice to handle double-escaped entities like &amp;amp;
    decoded = html.unescape(url)
    if "&" in decoded:
        decoded = html.unescape(decoded)

    return decoded


def normalize_subreddit(identifier: str) -> str:
    """
    Extract subreddit name from URL or identifier.

    Args:
        identifier: Subreddit identifier (can be URL, r/subreddit, or just subreddit)

    Returns:
        Normalized subreddit name
    """
    identifier = identifier.strip()

    # Match /r/name, r/name, or reddit.com/r/name
    # Allows for leading slash to be optional
    url_match = re.search(r"(?:reddit\.com)?/?r/(\w+)", identifier)
    if url_match:
        return url_match.group(1)

    # Remove r/ or /r/ prefix if still there (fallback)
    if identifier.startswith("/r/"):
        identifier = identifier[3:]
    elif identifier.startswith("r/"):
        identifier = identifier[2:]

    # Remove trailing slash and anything after it
    identifier = identifier.split("/")[0]

    # Remove anything after a colon or space if it looks like a label was accidentally included
    identifier = re.split(r"[:\s]", identifier)[0]

    return identifier


def extract_post_info_from_url(url: str) -> Dict[str, Optional[str]]:
    """
    Extract post ID and subreddit from Reddit URL.

    Format: https://reddit.com/r/{subreddit}/comments/{postId}/...

    Args:
        url: Reddit post URL

    Returns:
        Dict with 'subreddit' and 'post_id' keys
    """
    match = re.search(r"/r/(\w+)/comments/([a-zA-Z0-9]+)", url)
    if match:
        return {"subreddit": match.group(1), "post_id": match.group(2)}
    return {"subreddit": None, "post_id": None}


def validate_subreddit(subreddit: str) -> Dict[str, Any]:
    """
    Validate subreddit name.

    Args:
        subreddit: Subreddit name

    Returns:
        Dict with 'valid' (bool) and optional 'error' (str)
    """
    if not subreddit:
        return {"valid": False, "error": "Subreddit is required"}

    # Subreddit names: 2-21 characters, alphanumeric and underscores only
    if not re.match(r"^\w{2,21}$", subreddit):
        return {
            "valid": False,
            "error": "Invalid subreddit name. Use 2-21 alphanumeric characters or underscores.",
        }

    return {"valid": True}


def fetch_subreddit_info(subreddit: str, user_id: int) -> Dict[str, Optional[str]]:
    """
    Fetch subreddit information including icon using PRAW.

    Args:
        subreddit: Subreddit name (without /r/)
        user_id: User ID for authentication

    Returns:
        Dict with 'iconUrl' key
    """
    try:
        reddit = get_praw_instance(user_id)
        sub = reddit.subreddit(subreddit)

        # Try multiple fields for icon in order of preference
        # Access .icon_img first; PRAW lazily fetches subreddit data on attribute access
        raw_icon_url = sub.icon_img or sub.community_icon or getattr(sub, "header_img", None)

        icon_url = None
        if raw_icon_url:
            icon_url = fix_reddit_media_url(raw_icon_url)

        if icon_url:
            logger.debug(f"Fetched subreddit icon for r/{subreddit}: {icon_url}")

        return {"iconUrl": icon_url}

    except prawcore.exceptions.NotFound:
        logger.warning(f"Subreddit r/{subreddit} not found")
        return {"iconUrl": None}

    except prawcore.exceptions.Forbidden:
        logger.warning(f"Subreddit r/{subreddit} is private or banned")
        return {"iconUrl": None}

    except Exception as e:
        logger.warning(f"Failed to fetch subreddit info for r/{subreddit}: {e}")
        return {"iconUrl": None}


def extract_urls_from_text(text: str) -> list[str]:
    """
    Extract URLs from Reddit post text (selftext).

    Handles both plain URLs and markdown links [text](url).
    Decodes HTML entities in extracted URLs.

    Args:
        text: Text to extract URLs from

    Returns:
        List of URLs
    """
    if not text:
        return []

    urls = []

    # Pattern for markdown links: [text](url)
    markdown_link_pattern = re.compile(r"\[([^\]]*)\]\((https?://[^)]+)\)")
    for match in markdown_link_pattern.finditer(text):
        urls.append(decode_html_entities_in_url(match.group(2)))

    # Pattern for plain URLs: http:// or https://
    plain_url_pattern = re.compile(r"(?<!\]\()(https?://[^\s)]+)")
    for match in plain_url_pattern.finditer(text):
        # Remove trailing punctuation
        url = re.sub(r"[.,;:!?)]+$", "", match.group(1))
        decoded_url = decode_html_entities_in_url(url)
        if decoded_url not in urls:
            urls.append(decoded_url)

    return urls

File: src/lib/aggregators/sites/reddit/urls.ts
/**
 * Reddit URL utilities.
 *
 * Ported from old/core/aggregators/reddit/urls.py.
 */

/** The subset of `/r/{subreddit}/about.json`'s response `fetchSubredditInfo` reads. */
interface RedditSubredditAboutResponse {
  data?: {
    icon_img?: string;
    community_icon?: string;
    header_img?: string;
  };
}

export function decodeHtmlEntitiesInUrl(url: string): string {
  if (!url) return "";
  return (
    url
      .replace(/&amp;/g, "&")
      .replace(/&lt;/g, "<")
      .replace(/&gt;/g, ">")
      .replace(/&quot;/g, '"')
      .replace(/&apos;/g, "'")
      // `String.fromCodePoint` throws `RangeError` above 0x10FFFF (`&#1114112;`),
      // so a malformed ref must leave the matched text alone rather than take the
      // whole aggregation run down.
      .replace(/&#x([0-9a-fA-F]+);/g, (match, hex: string) => {
        try {
          return String.fromCodePoint(parseInt(hex, 16));
        } catch {
          return match;
        }
      })
      .replace(/&#(\d+);/g, (match, dec: string) => {
        try {
          return String.fromCodePoint(Number(dec));
        } catch {
          return match;
        }
      })
  );
}

export function fixRedditMediaUrl(url?: string | null): string | null {
  if (!url) return null;
  let decoded = decodeHtmlEntitiesInUrl(url);
  if (decoded.includes("&")) {
    decoded = decodeHtmlEntitiesInUrl(decoded);
  }
  return decoded;
}

export function normalizeSubreddit(identifier: string): string {
  if (!identifier) return "";
  let iden = identifier.trim();

  const match = iden.match(/(?:reddit\.com)?\/?r\/(\w+)/i);
  if (match && match[1]) {
    return match[1];
  }

  if (iden.startsWith("/r/")) {
    iden = iden.slice(3);
  } else if (iden.startsWith("r/")) {
    iden = iden.slice(2);
  }

  iden = iden.split("/")[0] || "";
  iden = iden.split(/[:\s]/)[0] || "";
  return iden;
}

export function extractPostInfoFromUrl(url: string): {
  subreddit: string | null;
  postId: string | null;
} {
  if (!url) return { subreddit: null, postId: null };
  const match = url.match(/\/r\/(\w+)\/comments\/([a-zA-Z0-9]+)/);
  if (match && match[1] && match[2]) {
    return { subreddit: match[1], postId: match[2] };
  }
  return { subreddit: null, postId: null };
}

export function validateSubreddit(subreddit: string): { valid: boolean; error?: string } {
  if (!subreddit) {
    return { valid: false, error: "Subreddit is required" };
  }

  if (!/^\w{2,21}$/.test(subreddit)) {
    return {
      valid: false,
      error: "Invalid subreddit name. Use 2-21 alphanumeric characters or underscores.",
    };
  }

  return { valid: true };
}

export async function fetchSubredditInfo(
  subreddit: string,
  _userId?: number | string | null,
  accessToken?: string | null,
): Promise<{ iconUrl: string | null }> {
  if (!subreddit) return { iconUrl: null };
  try {
    const url = accessToken
      ? `https://oauth.reddit.com/r/${subreddit}/about.json`
      : `https://www.reddit.com/r/${subreddit}/about.json`;
    const headers: Record<string, string> = { "User-Agent": "Yana/1.0" };
    if (accessToken) headers["Authorization"] = `Bearer ${accessToken}`;

    const res = await fetch(url, { headers, signal: AbortSignal.timeout(5000) });
    if (!res.ok) return { iconUrl: null };
    const data = (await res.json()) as RedditSubredditAboutResponse;
    const rawIcon =
      data?.data?.icon_img || data?.data?.community_icon || data?.data?.header_img || null;
    return { iconUrl: fixRedditMediaUrl(rawIcon) };
  } catch {
    return { iconUrl: null };
  }
}

export function extractUrlsFromText(text: string): string[] {
  if (!text) return [];
  const urls: string[] = [];

  const markdownLinkPattern = /\[([^\]]*)\]\((https?:\/\/[^)]+)\)/g;
  let match: RegExpExecArray | null;
  while ((match = markdownLinkPattern.exec(text)) !== null) {
    if (match[2]) {
      urls.push(decodeHtmlEntitiesInUrl(match[2]));
    }
  }

  const plainUrlPattern = /(?:^|[\s(])(https?:\/\/[^\s)]+)/g;
  while ((match = plainUrlPattern.exec(text)) !== null) {
    if (match[1]) {
      const cleanUrl = match[1].replace(/[.,;:!?)]+$/, "");
      const decoded = decodeHtmlEntitiesInUrl(cleanUrl);
      if (!urls.includes(decoded)) {
        urls.push(decoded);
      }
    }
  }

  return urls;
}
